src/Collector/VariableCollector.php

```diff
@@ -82,11 +86,16 @@ public function processNode(Node $node, Scope $scope): ?array
             return null;
         }
 
+        $variableType = $this->valueResolver->resolveAsConstantType($node->expr, $scope);
+        if ($variableType === null) {
```

Comment:
This can be directly done in TypeResolver, you send scope there

Reply:
It could but fallback to generic type from scope seems to me like VariableCollector specific behaviour that probably woukd not be wanted in other potential uses (there are none currently but I think they might be)

Reply:
Ok if we will create another use of it, we will see if scope fallback will be there too